diff --git a/tabbed.c b/tabbed.c
index ff3ada0..1f8b88b 100644
--- a/tabbed.c
+++ b/tabbed.c
@@ -1247,8 +1247,8 @@ xsettitle(Window w, const char *str)
 {
 	XTextProperty xtp;
 
-	if (XmbTextListToTextProperty(dpy, (char **)&str, 1,
-	    XCompoundTextStyle, &xtp) == Success) {
+	if (Xutf8TextListToTextProperty(dpy, (char **)&str, 1,
+	    XUTF8StringStyle, &xtp) == Success) {
 		XSetTextProperty(dpy, w, &xtp, wmatom[WMName]);
 		XSetTextProperty(dpy, w, &xtp, XA_WM_NAME);
 		XFree(xtp.value);
